src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -168,25 +276,44 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
                     // don't resolve builtins like `#[derive]`
                     if let Res::Def(..) = res {
                         let res = res.map_id(|_| panic!("unexpected node_id"));
-                        return Some(res);
+                        return Some(Ok(res));
                     }
                 }
             } else {
                 debug!("attempting to resolve item without parent module: {}", path_str);
+                return Some(Err(ResolutionFailure::NoParentItem));
             }
             None
         })
+        // This weird control flow is so we don't borrow the resolver more than once at a time
+        .unwrap_or_else(|| {
+            let mut split = path_str.rsplitn(2, "::");
+            if let Some((parent, base)) = split.next().and_then(|x| Some((split.next()?, x))) {
+                if let Some(res) = self.check_full_res(TypeNS, parent, parent_id, &None, &None) {
+                    return Err(if matches!(res, Res::PrimTy(_)) {
```

Comment:
```suggestion
                    // Check for `PrimTy` so we can distinguish `NoPrimitiveAssocItem`.
                    // `resolution_failure` needs this so it can give the name of the primitive;
                    // `prim.descr()` just says "builtin type".
                    return Err(if matches!(res, Res::PrimTy(_)) {
```